src/aktualizr_info/main.cc

```diff
@@ -72,32 +79,156 @@ int main(int argc, char **argv) {
+    }
+
+    if (vm.count("ecu-pub-key") != 0u) {
+      std::string key;
+      storage->loadPrimaryPublic(&key);
+      std::cout << key;
+      return EXIT_SUCCESS;
+    }
+
+    if (vm.count("ecu-prv-key") != 0u) {
+      std::string key;
+      storage->loadPrimaryPrivate(&key);
+      std::cout << key;
+      return EXIT_SUCCESS;
+    }
+
+    std::string director_root;
+    bool has_metadata = storage->loadLatestRoot(&director_root, Uptane::RepositoryType::Director());
+
+    // An arguments which depend on metadata.
+    std::string msg_metadata_fail = "Metadata is not available";
+    if (vm.count("images-root") != 0u) {
+      if (!has_metadata) {
+        std::cout << msg_metadata_fail << std::endl;
+      } else {
+        std::string images_root;
+        storage->loadLatestRoot(&images_root, Uptane::RepositoryType::Image());
+        std::cout << images_root << std::endl;
+      }
+      return EXIT_SUCCESS;
+    }
+
+    if (vm.count("images-target") != 0u) {
+      if (!has_metadata) {
+        std::cout << msg_metadata_fail << std::endl;
+      } else {
+        std::string images_targets;
+        storage->loadNonRoot(&images_targets, Uptane::RepositoryType::Image(), Uptane::Role::Targets());
+        std::cout << images_targets << std::endl;
+      }
+      return EXIT_SUCCESS;
+    }
+
+    if (vm.count("delegation") != 0u) {
+      if (!has_metadata) {
+        std::cout << msg_metadata_fail << std::endl;
+        return EXIT_SUCCESS;
+      } else {
+        return loadAndPrintDelegations(storage);
       }
-      std::cout << "Device ID: " << device_id << std::endl;
     }
 
+    if (vm.count("director-root") != 0u) {
+      if (!has_metadata) {
+        std::cout << msg_metadata_fail << std::endl;
+      } else {
+        std::cout << director_root << std::endl;
+      }
+      return EXIT_SUCCESS;
+    }
+
+    if (vm.count("director-target") != 0u) {
+      if (!has_metadata) {
+        std::cout << msg_metadata_fail << std::endl;
+      } else {
+        std::string director_targets;
+        storage->loadNonRoot(&director_targets, Uptane::RepositoryType::Director(), Uptane::Role::Targets());
+        std::cout << director_targets << std::endl;
+      }
+      return EXIT_SUCCESS;
+    }
+
+    if (vm.count("images-snapshot") != 0u) {
+      if (!has_metadata) {
+        std::cout << msg_metadata_fail << std::endl;
+      } else {
+        std::string snapshot;
+        storage->loadNonRoot(&snapshot, Uptane::RepositoryType::Image(), Uptane::Role::Snapshot());
+        std::cout << snapshot << std::endl;
+      }
+      return EXIT_SUCCESS;
+    }
+
+    if (vm.count("images-timestamp") != 0u) {
+      if (!has_metadata) {
+        std::cout << msg_metadata_fail << std::endl;
+      } else {
+        std::string timestamp;
+        storage->loadNonRoot(&timestamp, Uptane::RepositoryType::Image(), Uptane::Role::Timestamp());
+        std::cout << timestamp << std::endl;
+      }
+      return EXIT_SUCCESS;
+    }
+
```

Comment:
Regarding these multiple "if (<option-foo>) { do_foo(); return EXIT_SUCCESS;}", does it mean that it's not possible to get output of a few types of info/metadata within a single call to aktualizr-info ?

Also, these blocks of "if ... return EXIT_SUCCESS" statements look like ideal candidates for defining a map <std::string, std::function>  and having a single block of code that works for all options, e.g. "for op in options { if is_specified(op) { foo_map[op](); return EXIT_SUCCESS;}. This is rather matter of style and is minor comment that can be ignored.

Reply:
The early exit is probably undesirable. Whether the suggested refactoring is worth it is debatable; I have no strong feelings on that.

Reply:
> Regarding these multiple "if () { do_foo(); return EXIT_SUCCESS;}", does it mean that it's not possible to get output of a few types of info/metadata within a single call to aktualizr-info ?
> 
> Also, these blocks of "if ... return EXIT_SUCCESS" statements look like ideal candidates for defining a map <std::string, std::function> and having a single block of code that works for all options, e.g. "for op in options { if is_specified(op) { foo_mapop; return EXIT_SUCCESS;}. This is rather matter of style and is minor comment that can be ignored.



We do not need multiple outputs since we going to use the actualizr-info output in PIPE.

The block with "if .. return" looks ugly, but it does the same job you propose but generates less of CPU instructions.

Also, can we consider our main function as a single block of code? We can use exit() vs return but this particular case it's the same.

By the way, I using this tool https://godbolt.org/ to check the compiler output, time to time

Reply:
> We do not need multiple outputs since we going to use the actualizr-info output in PIPE.

I disagree. There is no reason not to support multiple outputs if the user asks for it. We've supported that until now, and I see no reason to remove it.

Reply:
Since we enable output for multiple arguments,  should we rename command --name-only?

Reply:
No, that is the one specific exception. I had forgotten about that; I made that as a special option for one of the backend devs. That one should indeed only print the name.